### PR TITLE
Make SSTable cleanup more efficient by fast forwarding to next owned range

### LINKED ISSUE #12998: Optionally use owned ranges in sstable reader to efficiently skip over ranges
Currently, cleanup compaction uses a `flat_mutation_reader_v2::filter` to skip over disowned partitions. This is inefficient.
So is the view update generator (See 10f8f13b9004fd5f73cab6a527be2fad291c044f).

https://github.com/scylladb/scylladb/pull/12422 allows cleaning up sstables implicitly in any compaction type
and it will also allow filtering the sstable for read from any path.
However, the filtering performance better be improved for that.
If we push the owned_ranges down to the reader, it could use the index to efficiently fast forward to the next owned token once it hits a disowned token.

### LINKED ISSUE #14317: Compactions that filter partitions may pollute index caching
Index caching is not scan resistant yet, so compactions that use index for forwarding to the next partition (e.g. cleanup), may end up polluting the index caching. Once the caching is scan resistant, MFU entries will remain cached, whereas LRU entries which aren't frequent will have lower priority over the most frequently used ones. We can easily fix the problem today by bypassing the cache with the help of partition slice's option set.

### PR DESCRIPTION
**Pull request by raphaelsc**
Today, SSTable cleanup skips to the next partition, one at a time, when it finds that the current partition is no longer owned by this node.

That's very inefficient because when a cluster is growing in size, existing nodes lose multiple sequential tokens in its owned ranges. Another inefficiency comes from fetching index pages spanning all unowned tokens, which was described in https://github.com/scylladb/scylladb/issues/14317.

To solve both problems, cleanup will now use multi range reader, to guarantee that it will only process the owned data and as a result skip unowned data. This results in cleanup scanning an owned range and then fast forwarding to the next one, until it's done with them all. This reduces significantly the amount of data in the index caching, as index will only be invoked at each range boundary instead.

Without further ado,

before:

`INFO  2023-07-01 07:10:26,281 [shard 0] compaction - [Cleanup keyspace2.standard1 701af580-17f7-11ee-8b85-a479a1a77573] Cleaned 1 sstables to [./tmp/1/keyspace2/standard1-b490ee20179f11ee9134afb16b3e10fd/me-3g7a_0s8o_06uww24drzrroaodpv-big-Data.db:level=0]. 2GB to 1GB (~50% of original) in 26248ms = 81MB/s. ~9443072 total partitions merged to 4750028.`

after:

`INFO  2023-07-01 07:07:52,354 [shard 0] compaction - [Cleanup keyspace2.standard1 199dff90-17f7-11ee-b592-b4f5d81717b9] Cleaned 1 sstables to [./tmp/1/keyspace2/standard1-b490ee20179f11ee9134afb16b3e10fd/me-3g7a_0s4m_5hehd2rejj8w15d2nt-big-Data.db:level=0]. 2GB to 1GB (~50% of original) in 17424ms = 123MB/s. ~9443072 total partitions merged to 4750028.`

Fixes #12998.
Fixes #14317.

### CODE REVIEW

**Comment by raphaelsc** [comment]
Sent in draft mode as I still have to polish a few things

**Comment by raphaelsc** [comment]
Had I used bigger partitions, we'd have seen a much better improvement than the one in the cover letter. As with smaller partitions, the today's filtering reader is able to skip to the next partition without invoking the index, as multiple partitions can fit in a single reader buffer.

File: compaction/compaction.cc [hunk 1]
```diff
@@ -788,7 +788,7 @@ class compaction {
         log_info("{} {} sstables to {}. {} to {} (~{}% of original) in {}ms = {}. ~{} total partitions merged to {}.",
                 report_finish_desc(),
                 _input_sstable_generations.size(), new_sstables_msg, utils::pretty_printed_data_size(_start_size), utils::pretty_printed_data_size(_end_size), int(ratio * 100),
-                std::chrono::duration_cast<std::chrono::milliseconds>(duration).count(), utils::pretty_printed_throughput(_end_size, duration),
+                std::chrono::duration_cast<std::chrono::milliseconds>(duration).count(), utils::pretty_printed_throughput(_start_size, duration),
```
Comment (bhalevy) [inline compaction/compaction.cc:820]: Let's ooen an issue for that as we may want to backport this patch to all live versions

Comment (raphaelsc) [reply compaction/compaction.cc:820]: Sure.

Comment (raphaelsc) [reply compaction/compaction.cc:820]: Done

**Comment by raphaelsc** [comment]
Ready for review.

Patch "sstables: Close SSTable reader if index exhaustion is detected in fast forward call" was added to make sure cleanup is able to incrementally release SSTables with its new multi range reader.

File: compaction/compaction.cc [hunk 2]
```diff
@@ -1456,8 +1481,14 @@ class scrub_compaction final : public regular_compaction {
...
-    flat_mutation_reader_v2 make_sstable_reader() const override {
-        auto crawling_reader = _compacting->make_crawling_reader(_schema, _permit, nullptr);
+    flat_mutation_reader_v2 make_sstable_reader(schema_ptr s,
+                                                reader_permit permit,
+                                                const dht::partition_range& range,
+                                                const query::partition_slice& slice,
+                                                tracing::trace_state_ptr trace,
+                                                streamed_mutation::forwarding sm_fwd,
+                                                mutation_reader::forwarding mr_fwd) const override {
+        auto crawling_reader = _compacting->make_crawling_reader(std::move(s), std::move(permit), nullptr);
```
Comment (tgrabiec) [inline compaction/compaction.cc:1507]: Here we ignore the restrictions, should we at least verify that they are for full range?

File: sstables/mx/reader.cc [hunk 3]
```diff
@@ -1700,7 +1718,9 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
             close_index_reader = _index_reader->close().finally([_ = std::move(_index_reader)] {});
         }
 
-        return when_all_succeed(std::move(close_context), std::move(close_index_reader)).discard_result().handle_exception([] (std::exception_ptr ep) {
+        auto sst = std::move(*this).detach_sstable();
```
Comment (tgrabiec) [inline sstables/mx/reader.cc:1721]: Above we manipulate _context directly, so here we could manipulate _sst directly too. detach_sstable() is not a public API, so &&-qualification seems unnecessary, and it's weird that only _sst is manipulated this way and _context is not.

**Review by tgrabiec — CHANGES_REQUESTED** [review]

File: compaction/compaction.cc (hunk 2 above)
Comment (raphaelsc) [reply compaction/compaction.cc:1507]: yes. we do something similar in sstable set's single key reader interface.

File: sstables/mx/reader.cc (hunk 3 above)
Comment (raphaelsc) [reply sstables/mx/reader.cc:1721]: Makes sense. I overcomplicated it. Better to just move sst into continuation.

**Comment by raphaelsc** [comment]
v2:
- Guarantee scrub works on full partition range
- Simplify detachment of SSTable in the reader once exhausted

**Review by tgrabiec — APPROVED** [review]

**Review by bhalevy — APPROVED** [review]
LGTM

**Comment by raphaelsc** [comment]
@scylladb/scylla-maint ping for merge

File: compaction/compaction.cc [hunk 4]
```diff
@@ -788,7 +788,7 @@ class compaction {
         log_info("{} {} sstables to {}. {} to {} (~{}% of original) in {}ms = {}. ~{} total partitions merged to {}.",
                 report_finish_desc(),
                 _input_sstable_generations.size(), new_sstables_msg, utils::pretty_printed_data_size(_start_size), utils::pretty_printed_data_size(_end_size), int(ratio * 100),
-                std::chrono::duration_cast<std::chrono::milliseconds>(duration).count(), utils::pretty_printed_throughput(_end_size, duration),
+                std::chrono::duration_cast<std::chrono::milliseconds>(duration).count(), utils::pretty_printed_throughput(_start_size, duration),
                 _cdata.total_partitions, _cdata.total_keys_written);
```
Comment (avikivity) [inline compaction/compaction.cc:822]: Please extract into a separate pull request (mainly so it can be featured in the newsletter).

File: compaction/compaction.cc [hunk 5]
```diff
@@ -631,17 +631,30 @@ class compaction {
...
 private:
     // Default range sstable reader that will only return mutation that belongs to current shard.
-    virtual flat_mutation_reader_v2 make_sstable_reader() const = 0;
+    virtual flat_mutation_reader_v2 make_sstable_reader(schema_ptr s,
+                                                        reader_permit permit,
+                                                        const dht::partition_range& range,
+                                                        const query::partition_slice& slice,
+                                                        tracing::trace_state_ptr,
+                                                        streamed_mutation::forwarding fwd,
+                                                        mutation_reader::forwarding) const = 0;
```
Comment (avikivity) [inline compaction/compaction.cc:630]: Can we keep the old interface (mostly because it's impossible to misuse) and instead add an interposer to skip?

Comment (avikivity) [reply compaction/compaction.cc:630]: An interposer will add its own slowdown too. Not sure.

File: sstables/mx/reader.cc [hunk 6]
```diff
@@ -1511,10 +1511,16 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
...
+    bool has_sstable_attached() const noexcept {
+        return bool(_sst);
+    }
     bool is_initialized() const {
         return bool(_context);
     }
     future<> initialize() {
+        if (!has_sstable_attached()) {
+            co_return;
+        }
```
Comment (avikivity) [inline sstables/mx/reader.cc:1523]: Please add comments here, it looks odd.

Another option is for the seek process to determine there's nothing in the range, and return an empty reader.

Comment (avikivity) [reply sstables/mx/reader.cc:1523]: Or maybe seek to the very end of the sstable and let it naturally terminate.

**Comment by raphaelsc** [comment]
> Looks very good! But the title should be less dramatic. The speedup will be around 1/(n+1) where n = nr_nodes/RF. It's indeed 50% (or actually 100%) for the initial doubling of the cluster capacity, but once the cluster is larger the ratio owned_ranges/previously_owned_ranges becomes closer to 1.

Adjusted the title. You're right, the improvement is proportional to the number of tokens extracted from existing nodes. With concurrent topology changes, we'd have better results after doubling the cluster capacity for example, but then we'll have tablets.

File: sstables/mx/reader.cc (hunk 6 above)
Comment (raphaelsc) [reply sstables/mx/reader.cc:1523]: If we enabled fast forward, combined reader cannot tell underlying readers are really exhausted.

Think that we're cleaning up a SSTable run, where each fragment overlaps a different owned range

When reading the first owned range, combined reader will call incremental selector and fetch the first fragment.

Once we're done exhausting it, underlying reader returns EOS

but is the reader really exhausted (I am asking this from combined reader's perspective)? It may have data for the next range we'll fast forward to, so it remains opened and listed by combined reader.

On the next fast forward call, the EOS flag of the underlying reader is cleared and the reader is asked again for data

Only the underlying reader itself knows that it has been exhausted, as it's the one that has the notion of EOF. 

The flat reader interface only knows about exhaustion of a reader for a particular range

File: compaction/compaction.cc (hunk 5 above)
Comment (denesb) [reply compaction/compaction.cc:630]: I think we should remove parameters that will never be used:
```
                                                        const query::partition_slice& slice,
                                                        tracing::trace_state_ptr,
                                                        streamed_mutation::forwarding fwd,
```

**Review by denesb — APPROVED** [review]
LGTM, one nit.

**Comment by raphaelsc** [comment]
v3:
- Added patch "sstables: Simplify sstable reader initialization"
- Documented why initialization doesn't take place when sstable was detached from reader

File: sstables/mx/reader.cc (hunk 6 above)
Comment (avikivity) [reply sstables/mx/reader.cc:1532]: I see. Then a mutation_source really is the best approach.

We could have hidden it in a layered reader, but it's also good to avoid adding cycles to the data path.

File: sstables/mx/reader.cc [hunk 7]
```diff
@@ -1523,7 +1527,7 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
 
             if (!present) {
                 _sst->get_filter_tracker().add_false_positive();
-                co_return;
+                co_return false;
```
Comment (avikivity) [inline sstables/mx/reader.cc:1541]: An even better interface is to return optional\<something>. It can be a method of sstable itself. nullopt means we don't need the something because we won't be reading anything. A something (reader? something used to initialize the reader?) can be used to actually to the reading.

But, that's for another day.

File: sstables/mx/reader.cc [hunk 8]
```diff
@@ -1591,8 +1590,8 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
             on_internal_error(sstlog, "mx reader: fast_forward_to(partition_range) not supported for reversed queries");
         }
 
-        return ensure_initialized().then([this, &pr] {
-            if (!is_initialized()) {
+        return maybe_initialize().then([this, &pr] (bool initialized) {
+            if (!initialized) {
```
Comment (avikivity) [inline sstables/mx/reader.cc:1605]: (if we get a nullopt we return an empty_reader rather than an sstable_reader that was primed to look empty)

File: sstables/mx/reader.cc [hunk 9]
```diff
@@ -1612,6 +1623,15 @@ class mx_sstable_mutation_reader : public mp_row_consumer_reader_mx {
...
                     _read_enabled = false;
+                    if (_index_reader->eof()) {
+                        // Close the SSTable reader proactively, if the index is completely exhausted
+                        // and no partition was found in the current fast-forward call. This allows
+                        // disk space of SSTables to be reclaimed earlier if they take part in a
+                        // long-living read and they're deleted midway.
+                        sstlog.trace("Closing reader {} for {} after fast-forward call found that index reached EOF and there's nothing left to read",
+                                     fmt::ptr(this), _sst->get_filename());
+                        return close();
+                    }
```
Comment (avikivity) [inline sstables/mx/reader.cc:1634]: This really frightens me. Let's see what tests show.

Comment (raphaelsc) [reply sstables/mx/reader.cc:1634]: I am convinced it works, but I'd also like to have something that will fit more naturally with the flow. A reader is only closed once destroyed by whoever keeps it. I share your feeling and have been giving this some thought.

I wouldn't like to penalize others, with always having an additional layer of indirection.

So what about wrapping it only for cleanup with a 

auto_closed_sstable_reader, which stores the sstable partition range and the sstable reader itself. any request is forwarded to the underlying reader

if fast forward is called beyond the range spanned by the sstable, we can safely close the underlying sstable reader and replace it with an empty reader.

Much better, right?

Comment (raphaelsc) [reply sstables/mx/reader.cc:1634]: I think it looks good: https://github.com/raphaelsc/scylla/commit/a94ecd422f2b470c977b1efba7320932d468cbf3#diff-79da4c6195a1867e40a679218fb17f5e308e433983d351da9adbdcf32271e341R1227
